**What was tried:** QA workflow run [#25700722340](https://github.com/svv2014/loop/actions/runs/25700722340) on PR #326 (branch `fix/issue-301-scanner-atomic-lock`).

**What failed:**
```
No structured test names found
```

<details><summary>Last 20 lines of failure output</summary>

```
qa-merge	Run tests	2026-05-11T22:22:10.6382934Z ok 513 validate: duplicate trigger_label within issue_stages fails
qa-merge	Run tests	2026-05-11T22:22:10.6974468Z ok 514 validate: same trigger_label across issue_stages and pr_stages is allowed
qa-merge	Run tests	2026-05-11T22:22:10.7570831Z ok 515 validate: unknown transition target fails
qa-merge	Run tests	2026-05-11T22:22:10.8169092Z ok 516 validate: transition to terminal label 'done' passes
qa-merge	Run tests	2026-05-11T22:22:10.8771337Z ok 517 validate: decisions approve/reject referencing unknown labels fails
qa-merge	Run tests	2026-05-11T22:22:10.9362695Z ok 518 validate: empty workflow with no stages fails
qa-merge	Run tests	2026-05-11T22:22:10.9957318Z ok 519 validate: stage missing trigger_label fails
qa-merge	Run tests	2026-05-11T22:22:11.0554501Z ok 520 validate: stage missing handler fails
qa-merge	Run tests	2026-05-11T22:22:11.1183312Z ok 521 validate: missing handler script produces warning but passes
qa-merge	Run tests	2026-05-11T22:22:11.2891682Z ok 522 default.yaml: every referenced label is canonical
qa-merge	Run tests	2026-05-11T22:22:11.3608509Z ok 523 minimal.yaml: every referenced label is canonical
qa-merge	Run tests	2026-05-11T22:22:11.4522813Z ok 524 docs-only.yaml: every referenced label is canonical
qa-merge	Run tests	2026-05-11T22:22:11.5906825Z ok 525 all shipped workflows pass loop_workflow_validate
qa-merge	Run tests	2026-05-11T22:22:11.8217893Z ok 526 no deprecated alias appears as a workflow label
qa-merge	Run tests	2026-05-11T22:22:11.8867151Z ok 527 no-op when WORKTREE_EXTRA_PATHS is unset
qa-merge	Run tests	2026-05-11T22:22:11.9214794Z ok 528 symlinks each entry from primary into worktree
qa-merge	Run tests	2026-05-11T22:22:11.9462930Z ok 529 skips entries that don't exist in primary
qa-merge	Run tests	2026-05-11T22:22:11.9747020Z ok 530 skips entries that already exist in worktree
qa-merge	Run tests	2026-05-11T22:22:12.0057856Z ok 531 creates parent directories before linking
qa-merge	Run tests	2026-05-11T22:22:12.0492827Z ##[error]Process completed with exit code 1.
```
</details>

**Options:**
(a) Fix the failing test(s) — update the code or test to make them pass, then push a new commit to this PR.
(b) Skip the test(s) — if the test is flaky or irrelevant to this change, annotate it with the appropriate skip marker and document why.

--- a/lib/recovery.sh
+++ b/lib/recovery.sh
@@ -239,9 +239,11 @@ recovery_check_stuck_labels() {
         op_label="${op_entry%%:*}"
         trigger_canon="${op_entry#*:}"
 
-        # Resolve the project-specific label name for both labels
+        # Operational labels (in-progress, in-rework, in-review) are set
+        # verbatim by handlers — do not translate them via loop_label_for.
+        # Only the trigger to restore needs project-specific resolution.
         local actual_op actual_trigger
-        actual_op=$(loop_label_for "$slug" "$op_label" 2>/dev/null || echo "$op_label")
+        actual_op="$op_label"
         actual_trigger=$(loop_label_for "$slug" "$trigger_canon" 2>/dev/null || echo "$trigger_canon")
 
         # Fetch issues with this operational label

--- a/scanner/reconciler.sh
+++ b/scanner/reconciler.sh
@@ -660,7 +660,7 @@ PY
         local repo_slug="${repo//\//-}"
         local sentinel="$cool_down_dir/${repo_slug}-${num}"
         if [ -f "$sentinel" ]; then
-            local mtime; mtime=$(stat -f %m "$sentinel" 2>/dev/null || stat -c %Y "$sentinel" 2>/dev/null || echo 0)
+            local mtime; mtime=$(stat -c %Y "$sentinel" 2>/dev/null || stat -f %m "$sentinel" 2>/dev/null || echo 0)
             local age=$(( now_epoch - mtime ))
             if [ "$age" -lt "$cool_down_seconds" ]; then
                 log "[$repo] LOST issue #$num — Signal cooled-down (last notified ${age}s ago)"
@@ -1597,7 +1597,7 @@ PY
         local now_epoch; now_epoch=$(date +%s)
 
         if [ -f "$sentinel" ]; then
-            local mtime; mtime=$(stat -f %m "$sentinel" 2>/dev/null || stat -c %Y "$sentinel" 2>/dev/null || echo 0)
+            local mtime; mtime=$(stat -c %Y "$sentinel" 2>/dev/null || stat -f %m "$sentinel" 2>/dev/null || echo 0)
             local age=$(( now_epoch - mtime ))
             if [ "$age" -lt "$cool_down_seconds" ]; then
                 log "[$repo] anomaly: #$num touches=$touches in ${window_hours}h — Signal cooled-down (${age}s ago)"
@@ -1705,7 +1705,7 @@ for c in data.get("comments", [])[-10:]:
         local repo_slug="${repo//\//-}"
         local sentinel="$state_dir/${repo_slug}-${num}"
         if [ -f "$sentinel" ]; then
-            local mtime; mtime=$(stat -f %m "$sentinel" 2>/dev/null || stat -c %Y "$sentinel" 2>/dev/null || echo 0)
+            local mtime; mtime=$(stat -c %Y "$sentinel" 2>/dev/null || stat -f %m "$sentinel" 2>/dev/null || echo 0)
             local age=$(( now_epoch - mtime ))
             if [ "$age" -lt "$cool_down_seconds" ]; then
                 log "[$repo] agent-distress: #$num matched '$matched' — Signal cooled-down (${age}s ago)"

--- a/scanner/scanner.sh
+++ b/scanner/scanner.sh
@@ -114,16 +114,27 @@ _budget_exceeded() {
 }
 
 acquire_lock() {
-    if [ -f "$LOCK_FILE" ]; then
-        local pid
-        pid=$(cat "$LOCK_FILE" 2>/dev/null || echo "")
-        if [ -n "$pid" ] && kill -0 "$pid" 2>/dev/null; then
-            log "Already running (PID $pid). Exiting."
-            exit 0
-        fi
+    # Atomic create — POSIX noclobber fails if the file already exists.
+    if ( set -o noclobber; echo "$$" > "$LOCK_FILE" ) 2>/dev/null; then
+        trap 'rm -f "$LOCK_FILE"' EXIT
+        return 0
     fi
-    echo $$ > "$LOCK_FILE"
-    trap 'rm -f "$LOCK_FILE"' EXIT
+    # File already exists — check if the holder is still alive.
+    local pid
+    pid=$(cat "$LOCK_FILE" 2>/dev/null || echo "")
+    if [ -n "$pid" ] && kill -0 "$pid" 2>/dev/null; then
+        log "Already running (PID $pid). Exiting."
+        exit 0
+    fi
+    # Stale lock — holder PID is dead; remove and retry once.
+    rm -f "$LOCK_FILE"
+    if ( set -o noclobber; echo "$$" > "$LOCK_FILE" ) 2>/dev/null; then
+        trap 'rm -f "$LOCK_FILE"' EXIT
+        return 0
+    fi
+    # Another process won the race on the retry — exit cleanly.
+    log "Lock unavailable after stale-PID recovery. Exiting."
+    exit 0
 }
 
 # Dedup cache — prevent emitting the same event every tick.

--- a/tests/handlers.bats
+++ b/tests/handlers.bats
@@ -49,15 +49,15 @@ teardown() {
     backend_remove_label "owner/test-repo" "1" "$_dev_trigger"
     backend_add_label    "owner/test-repo" "1" "in-progress"
 
-    # trigger label must be removed
-    grep -q "remove dev" "$ops_log"
+    # trigger label must be removed (default workflow uses needs-dev)
+    grep -q "remove needs-dev" "$ops_log"
     # in-progress must be added
     grep -q "add in-progress" "$ops_log"
     # trigger label must NOT be re-added (no dual state)
-    ! grep -q "add dev" "$ops_log"
+    ! grep -q "add needs-dev" "$ops_log"
     # remove must precede add
     local rm_line add_line
-    rm_line=$(grep -n "remove dev" "$ops_log" | cut -d: -f1)
+    rm_line=$(grep -n "remove needs-dev" "$ops_log" | cut -d: -f1)
     add_line=$(grep -n "add in-progress" "$ops_log" | cut -d: -f1)
     [ "$rm_line" -lt "$add_line" ]
 }
@@ -110,11 +110,12 @@ YAML
     backend_remove_label "owner/test-repo" "1" "$_po_trigger"
     backend_add_label    "owner/test-repo" "1" "in-progress"
 
-    grep -q "remove po-review" "$ops_log"
+    # default workflow po trigger is needs-po
+    grep -q "remove needs-po" "$ops_log"
     grep -q "add in-progress" "$ops_log"
-    ! grep -q "add po-review" "$ops_log"
+    ! grep -q "add needs-po" "$ops_log"
     local rm_line add_line
-    rm_line=$(grep -n "remove po-review" "$ops_log" | cut -d: -f1)
+    rm_line=$(grep -n "remove needs-po" "$ops_log" | cut -d: -f1)
     add_line=$(grep -n "add in-progress" "$ops_log" | cut -d: -f1)
     [ "$rm_line" -lt "$add_line" ]
 }
@@ -278,7 +279,8 @@ EOF
     _QA_LABEL=$(loop_label_for "test-proj" "needs-qa")
 
     [ "$_REVIEW_LABEL" = "needs-review" ]
-    [ "$_REWORK_LABEL" = "needs-rework" ]
+    # default workflow rework stage uses needs-dev as its trigger_label
+    [ "$_REWORK_LABEL" = "needs-dev" ]
     [ "$_QA_LABEL" = "needs-qa" ]
 
     local SLUG=test-proj REPO=owner/test-repo ISSUE_NUM=1
@@ -292,9 +294,9 @@ IMPORTANT: label '${_REVIEW_LABEL}' (or 'needs-clarification' if blocked).
 EOF
     )
 
-    # Default workflow: canonical names must appear in the prompt
+    # Default workflow: resolved trigger names must appear in the prompt
     [[ "$prompt" == *"needs-review"* ]]
-    [[ "$prompt" == *"needs-rework"* ]]
+    [[ "$prompt" == *"needs-dev"* ]]
     [[ "$prompt" == *"needs-qa"* ]]
     [[ "$prompt" != *"review-pending"* ]]
 }

--- a/tests/reconciler-alias-rename.bats
+++ b/tests/reconciler-alias-rename.bats
@@ -31,6 +31,20 @@ setup() {
         local var="MOCK_ISSUES_${lbl//-/_}"
         eval "echo \"\${$var:-[]}\""
     }
+    # _reconcile_label_renames calls backend_list_issues_with_label and
+    # backend_list_prs_with_label (NDJSON: one object per line).
+    # Both must filter by the requested label so canonical tickets are untouched.
+    backend_list_issues_with_label() {
+        local _repo="$1" lbl="$2"
+        local var="MOCK_ISSUES_${lbl//-/_}"
+        local arr; eval "arr=\"\${$var:-[]}\""
+        echo "$arr" | jq -c '.[]' 2>/dev/null || true
+    }
+    backend_list_prs_with_label() {
+        local _repo="$1" lbl="$2"
+        echo "${MOCK_PRS_JSON:-[]}" | \
+            jq -c --arg lbl "$lbl" '.[] | select(.labels[] | .name == $lbl)' 2>/dev/null || true
+    }
     backend_add_label()    { echo "add_label $2 $3"    >> "$OPS_LOG"; }
     backend_remove_label() { echo "remove_label $2 $3" >> "$OPS_LOG"; }
     log() { :; }

--- a/tests/reconciler.bats
+++ b/tests/reconciler.bats
@@ -12,7 +12,7 @@ setup() {
     export LOOP_ROOT="$REPO_ROOT"
     export LOOP_WORKFLOW_DIR="$REPO_ROOT/config/workflows"
 
-    # Minimal project config so loop_label_for resolves to "dev".
+    # Minimal project config so loop_label_for resolves to "needs-dev".
     cat > "$BATS_TMPDIR/fixture.yaml" <<'YAML'
 version: 1
 projects:
@@ -108,8 +108,8 @@ teardown() {
     # blocked must be removed.
     grep -q "remove_label 10 blocked" "$OPS_LOG"
 
-    # trigger label (dev) must be added.
-    grep -q "add_label 10 dev" "$OPS_LOG"
+    # trigger label (needs-dev in default workflow) must be added.
+    grep -q "add_label 10 needs-dev" "$OPS_LOG"
 
     # comment must be posted on the issue.
     grep -q "comment_issue 10" "$OPS_LOG"
@@ -130,8 +130,8 @@ teardown() {
     [ "$status" -eq 0 ]
 
     local rm_line add_line
-    rm_line=$(grep -n "remove_label 20 blocked" "$OPS_LOG" | cut -d: -f1)
-    add_line=$(grep -n "add_label 20 dev"       "$OPS_LOG" | cut -d: -f1)
+    rm_line=$(grep -n "remove_label 20 blocked"   "$OPS_LOG" | cut -d: -f1)
+    add_line=$(grep -n "add_label 20 needs-dev"   "$OPS_LOG" | cut -d: -f1)
     [ "$rm_line" -lt "$add_line" ]
 }
 
@@ -237,7 +237,7 @@ print(t.strftime('%Y-%m-%dT%H:%M:%SZ'))
     run recovery_check_stuck_labels "test-proj"
     [ "$status" -eq 0 ]
     grep -q "remove_label 42 in-progress" "$BATS_TMPDIR/ops.log"
-    grep -q "add_label 42 dev"            "$BATS_TMPDIR/ops.log"
+    grep -q "add_label 42 needs-dev"      "$BATS_TMPDIR/ops.log"
 }
 
 @test "recovery_check_stuck_labels: skips issue with live handler lock" {
@@ -279,7 +279,8 @@ print(t.strftime('%Y-%m-%dT%H:%M:%SZ'))
     run recovery_check_stuck_labels "test-proj"
     [ "$status" -eq 0 ]
     grep -q "remove_label 55 in-rework"   "$BATS_TMPDIR/ops.log"
-    grep -q "add_label 55 needs-rework"   "$BATS_TMPDIR/ops.log"
+    # rework stage trigger in default workflow is needs-dev
+    grep -q "add_label 55 needs-dev"      "$BATS_TMPDIR/ops.log"
 }
 
 @test "recovery_check_stuck_labels: dry-run suppresses label mutations" {

--- a/tests/scanner.bats
+++ b/tests/scanner.bats
@@ -657,3 +657,29 @@ YAML
     after_count=$(find "$DEDUP_DIR" -type f | wc -l)
     [ "$before_count" -eq "$after_count" ]
 }
+
+# ---------------------------------------------------------------------------
+# acquire_lock — atomic noclobber locking
+# ---------------------------------------------------------------------------
+
+# Verify that a second scanner invocation exits immediately (without emitting)
+# when the lock file already holds a live PID.
+# Strategy: invoke scanner.sh --dry-run with a pre-written lock file whose PID
+# is our current test process ($$) — guaranteed alive. The scanner must detect
+# the live holder and exit 0 without printing any "DRY-RUN emit:" lines.
+@test "acquire_lock: exits 0 without dispatching when live PID holds the lock" {
+    local lock_file="/tmp/loop-scanner.lock"
+    # Write our own PID as the holder — it is guaranteed to be alive.
+    echo "$$" > "$lock_file"
+
+    # Run the full scanner script in --dry-run mode. It should detect the live
+    # PID and exit 0 without emitting anything.
+    run "$REPO_ROOT/scanner/scanner.sh" --dry-run
+
+    # Always clean up the lock file we created.
+    rm -f "$lock_file"
+
+    [ "$status" -eq 0 ]
+    # Confirm no events were dispatched.
+    [[ "$output" != *"DRY-RUN emit:"* ]]
+}

